### PR TITLE
Update cart/checkout usage of the @wordpress/components Slot Fill

### PR DESCRIPTION
**Pull request by samueljseay**
#### Rationale

The goal of #45571 epic is to replace our outdated usage of a very old version of `@wordpress/components` in the cart and checkout blocks.

In my opinion it is not worth trying to replace the `@wordpress/components` Slot Fill implementation as part of this, for several reasons:

1. The rationale that wp components is not suited to front-ends doesn't apply to the slot fill, its not styled or user facing behaviour. It's more like a utility for React.

2. Once wp components is tree-shaken to just the slot fill library it will be very small

3. There is not a really good drop-in replacement that would give backwards compat, not a good use of our time to reimplement slot fill functionality, even if a library comes close

4. Finally, we're looking in future to provide more of these extension points via the block hooks API. With that in mind it doesn't make much sense to invest further in a Slot Fill implementation.

#### Approach

~My approach here still leaves in the place the old `wordpress-components` alias because I don't want to break the deprecated filter blocks that consume `FormTokenField`, but later when we remove those blocks it can be removed altogether.~

~Instead what I've done here is bundle the latest `@wordpress/components` not as an external for the front-end bundle. Webpack will tree-shake out everything that is not related to SlotFill.~

I have changed the approach, because my first implementation was naive, and relying on import `@wordpress/components` will likely cause conflicts and issues due to the import being handled by the Dependency Extraction Webpack Plugin. 

Here's what I've done instead:

1. Add another wordpress components alias. I've pegged it to WP6.5 because some of the imports in it are not compliant with our existing imports (e.g. functions that are not yet exposed in our current import of `@wordpress/compose`

2. I've replaced the old alias imports but I've imported directly from the slot fill directory in the build output, this avoids some conflict issues when importing the index of `@wordpress/components` more than once. It also minimizes any side effects being run that we don't need.

Note: I would have loved to avoid another alias to wp components, but it's not possible here, and I think it's ok because the surface area we've exposed is very small. I will update documentation to ensure that others don't use this alias to import any other part of the package.

#### Note on bundle size

1.  The old wordpress components  (14.2.0) package still includes SlotFill for it's internal dependencies. e.g. Popover uses SlotFill. We will only be able to remove that when we remove the dependency on the old version altogether.

2. Adding just slot-fill from the new version (wp-6.5) adds 3kb gzipped. So not really a lot.

Screenshot 2024-05-03 at 12 44 06 PM

#### Extensions using the Blocks Slot/Fills

1. Install WooCommerce Subscriptions and create a Subscription product.
2. Set up two shipping methods (e.g. Free and Flat rate)
3. Add this product to your cart and go to the Cart block.
4. Ensure you see a "Recurring total" section in the sidebar, it should look similar to this
image

5. Ensure you see an "Initial shipment" and "Recurring shipment" option in the sidebar, they should look like this
image

7. Go to the Checkout block and ensure the same totals section from step 4 is present.
8. Go to the Shipping Options section and ensure it shows a recurring and initial shipment showing all shipping options, like so:
<image src="https://github.com/woocommerce/woocommerce/assets/5656702/0b1d95d5-95c7-4cf8-8121-297c16b73d87" width="300" />

#### Extensions with `@wordpress/components` as a dependency

- Install the "Shipping workshop" plugin 
[shipping-workshop.zip](https://github.com/woocommerce/woocommerce/files/15272804/shipping-workshop.zip) and activate it.
- Throughout the process, have your JS console open and be sure to look out for any errors relating to `@wordpress/components` or other JS packages not being found.
- Ensure your store has shipping methods enabled. Disable local pickup.
- Add an item to your cart and go to the Checkout block.
- Ensure there's a select in the shipping methods section for `IF I AM NOT AT HOME PLEASE…`
- Interact with it and choose the "Other" option - ensure a text field renders and you can type in it.
- Type in it and check out.
- Ensure the details you entered are visible on the order confirmation screen.

image

Order confirmation:
image

#### Dev testing

The simplest way to test this is to create a plugin using create block and test that a slot fill still renders on this branch the same way it does on trunk:

1.  run `npx @wordpress/create-block -t @woocommerce/extend-cart-checkout-block your-extension-name`
10. Build (`npm build`) and mount the newly created plugin in your dev environment. ensure the plugin is activated
11. Go to the cart and or checkout pages 

See that the slotfill is rendered like below:
Screenshot 2024-05-02 at 5 29 41 PM

Please also test the shipping tutorial extension.

Testing instructions outlined in this comment: https://github.com/woocommerce/woocommerce/pull/47105#issuecomment-2104235466 

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [x] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 
Update cart/checkout usage of the @wordpress/components Slot Fill

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce-blocks/bin/webpack-configs.js
Comment (samueljseay) [inline plugins/woocommerce-blocks/bin/webpack-configs.js]: These overrides only apply to the front-end bundle.

**Comment by nielslange** [comment]
@samueljseay There's one failing e2e test:
[`[blockThemeWithGlobalSideEffects] › mini-cart/mini-cart-block.merchant.block_theme.side_effects.spec.ts:42:7 › Merchant → Mini Cart › in FSE editor › can only be inserted once`](https://github.com/woocommerce/woocommerce/actions/runs/8934265445/job/24582709202?pr=47105)

@Aljullu fixed this issue a last week with https://github.com/woocommerce/woocommerce/pull/47134. If you merge `trunk` into your PR, this particular test should pass.

File: plugins/woocommerce-blocks/assets/js/base/README.MD
Comment (nielslange) [inline plugins/woocommerce-blocks/assets/js/base/README.MD]: > _"[...] please use Ariakit to build them."_

What are your thoughts on adding a link to the corresponding page?

File: plugins/woocommerce-blocks/bin/webpack-configs.js
Comment (nielslange) [inline plugins/woocommerce-blocks/bin/webpack-configs.js]: Shouldn't _"dewp"_ be _"dep"_, given that you're referring to the dependency handlers?

**Review by nielslange — COMMENTED** [review]
Thanks for working on this, @samueljseay. I've successfully tested the functionality. Regarding the Webpack changes, I'd appreciate it if @opr could give it a second look. While they look good to me, I’m not yet an expert in this area and would appreciate a second opinion.

File: plugins/woocommerce-blocks/assets/js/base/README.MD
Comment (samueljseay) [reply plugins/woocommerce-blocks/assets/js/base/README.MD]: Great idea and an easy fix. will do

File: plugins/woocommerce-blocks/bin/webpack-configs.js
Comment (samueljseay) [reply plugins/woocommerce-blocks/bin/webpack-configs.js]: @nielslange I'm referring to **D**ependency **E**xtraction **W**ebpack **P**lugin

but it's not clear so I'll just type the whole thing here, because probably not everyone knows what that is.

Comment (samueljseay) [reply plugins/woocommerce-blocks/bin/webpack-configs.js]: Fixed in https://github.com/woocommerce/woocommerce/pull/47105/commits/3c71600dc6d8cf66cd88451ec3d2957be3b249c5

File: plugins/woocommerce-blocks/assets/js/base/README.MD
Comment (samueljseay) [reply plugins/woocommerce-blocks/assets/js/base/README.MD]: Addressed in https://github.com/woocommerce/woocommerce/pull/47105/commits/e6aac166ec8d39352a8f6c42d402be4d254954ae

Comment (nielslange) [reply plugins/woocommerce-blocks/assets/js/base/README.MD]: Nice! Thanks for considering this, @samueljseay. 🙌

File: plugins/woocommerce-blocks/assets/js/base/README.MD
Comment (ralucaStan) [inline plugins/woocommerce-blocks/assets/js/base/README.MD]: Let's mention that we already have this library and give one example of a current usecase so that there is a starting point for people looking to work on a primitive component: `import { Button as AriakitButton } from '@ariakit/react';`

File: plugins/woocommerce-blocks/bin/webpack-configs.js
Comment (ralucaStan) [inline plugins/woocommerce-blocks/bin/webpack-configs.js]: could you expand your comment here? when would that be needed?

File: plugins/woocommerce-blocks/bin/webpack-configs.js
Comment (ralucaStan) [inline plugins/woocommerce-blocks/bin/webpack-configs.js]: why is this line needed?

File: plugins/woocommerce-blocks/package.json
Comment (ralucaStan) [inline plugins/woocommerce-blocks/package.json]: 😮

File: plugins/woocommerce-blocks/tests/e2e/tests/checkout/checkout-block.shopper.block_theme.side_effects.spec.ts
Comment (ralucaStan) [inline plugins/woocommerce-blocks/tests/e2e/tests/checkout/checkout-block.shopper.block_theme.side_effects.spec.ts]: is this change related to the PR?

File: plugins/woocommerce-blocks/assets/js/base/README.MD
Comment (ralucaStan) [reply plugins/woocommerce-blocks/assets/js/base/README.MD]: I also recommend updating this part of the documentation; https://github.com/woocommerce/woocommerce/blob/f040e3acf7df9420a09d37b84358ac7d2e03b8a3/plugins/woocommerce-blocks/docs/contributors/javascript-build-system.md?plain=1#L50

it still mentions using wordpress-components on the front-end

**Review by ralucaStan — COMMENTED** [review]
I added some comments to this PR, to gain more clarity and help with the docs updates. Thank you Sam

**Review by opr — CHANGES_REQUESTED** [review]
@samueljseay - thanks for your work on this, after applying these changes, one of the example plugins we use to showcase extensibility options (https://github.com/woocommerce/wceu23-shipping-workshop-final) stops working. Specifically the error is:

image

`Error: Cannot read properties of undefined (reading 'SelectControl')`

The problem is `wp.components` is no longer on the window. The plugin itself does this:

`import { SelectControl, TextareaControl } from '@wordpress/components';`.

and it is using `@woocommerce/dependency-extraction-webpack-plugin`.

`wp-components` is listed as a dependency of the plugin's script, and is being loaded. 

I'd expect this to still be available on the window if specifically importing `@wordpress/components` in another plugin, what are your thoughts on this?

Removing the changes to `webpack-configs.js` fixes this by the way. I'm wondering why _our_ webpack config would completely prevent this going on the window 🤔

File: plugins/woocommerce-blocks/assets/js/base/README.MD
Comment (samueljseay) [reply plugins/woocommerce-blocks/assets/js/base/README.MD]: @ralucaStan That is still relevant since wordpress components exists. But maybe we could add an advisement that it's deprecated and not to use it?

File: plugins/woocommerce-blocks/bin/webpack-configs.js
Comment (samueljseay) [reply plugins/woocommerce-blocks/bin/webpack-configs.js]: @ralucaStan I'll try add a bit more explanation, but maybe it'd be best to point to the docs for that plugin. I'll come up with something.

**Comment by samueljseay** [comment]
@opr thank you for testing that

> I'd expect this to still be available on the window if specifically importing @wordpress/components in another plugin, what are your thoughts on this?

Yes this is curious and concerning. Their instance of the plugin should still cause the dependency to be extracted to the window as normal and their dewp handler should load from the window while ours skips it just for our front-end bundle, so I'm going to have to investigate this, obviously we don't want to break that for other developers.

Also curiously these are editor side dependencies so even stranger.

File: plugins/woocommerce-blocks/bin/webpack-configs.js
Comment (samueljseay) [reply plugins/woocommerce-blocks/bin/webpack-configs.js]: This line allows the dewp to extract deps to the window, but now I'm blanking if this is actually needed in this circumstance. I'll double check.

File: plugins/woocommerce-blocks/tests/e2e/tests/checkout/checkout-block.shopper.block_theme.side_effects.spec.ts
Comment (samueljseay) [reply plugins/woocommerce-blocks/tests/e2e/tests/checkout/checkout-block.shopper.block_theme.side_effects.spec.ts]: Hmm, I think I have mistakenly pushed this e2e fix from another branch to this one 😅 Thanks I'll see if I can revert this.

**Comment by opr** [comment]
@samueljseay please can we add testing with [woocommerce/wceu23-shipping-workshop-final](https://github.com/woocommerce/wceu23-shipping-workshop-final) as part of the testing steps for this PR?

**Comment by samueljseay** [comment]
@opr we absolutely can, but I am going to need more detailed testing instructions.

I have taken a different approach and it doesn't cause exceptions to be thrown in the plugin, but I don't know how to confirm this example plugin is working.

I will fill in all the details of this new approach and address existing feedback that still applies but I just need some support in the meantime to ensure that we test the right things here.

File: plugins/woocommerce-blocks/bin/webpack-configs.js
Comment (samueljseay) [reply plugins/woocommerce-blocks/bin/webpack-configs.js]: No longer relevant.

File: plugins/woocommerce-blocks/package.json
Comment (samueljseay) [inline plugins/woocommerce-blocks/package.json:319]: The latest version was a bit new, it imports some dependencies that we already externalize like wordpress/compose that have new exposed interfaces we don't have in our version yet.

File: plugins/woocommerce-blocks/packages/checkout/slot/index.js
Comment (samueljseay) [inline plugins/woocommerce-blocks/packages/checkout/slot/index.js:19]: This stuff doesn't make such sense when we peg the version of slot fill that we import. maybe we pegged the version after we added this code, i don't know. Either way we can be confident this API won't change from under us. We are also no longer importing experimental interfaces exposed from the index of wp components.

**Comment by samueljseay** [comment]
@opr requested re-review. Lets collaborate on some testing instructions for that example plugin when you get a chance. I've updated the description to better reflect the new changes.

**Comment by opr** [comment]
Thanks a lot! Will review again soon.

> Lets collaborate on some testing instructions for that example plugin when you get a chance. I've updated the description to better reflect the new changes.

I'd recommend the following:

- Install the "Shipping workshop" plugin 
[shipping-workshop.zip](https://github.com/woocommerce/woocommerce/files/15272804/shipping-workshop.zip) and activate it.
- Throughout the process, have your JS console open and be sure to look out for any errors relating to `@wordpress/components` or other JS packages not being found.
- Ensure your store has shipping methods enabled. Disable local pickup.
- Add an item to your cart and go to the Checkout block.
- Ensure there's a select in the shipping methods section for `IF I AM NOT AT HOME PLEASE…`
- Interact with it and choose the "Other" option - ensure a text field renders and you can type in it.
- Type in it and check out.
- Ensure the details you entered are visible on the order confirmation screen.

image

Order confirmation:
image

**Comment by samueljseay** [comment]
@opr I have tested the extension as you outlined and it works great for me. I've added a link to your testing instructions in the PR description.

**Comment by samueljseay** [comment]
I would still really appreciate another person to look at testing the extension here, but just noting that fixing the tests is blocked by merging https://github.com/woocommerce/woocommerce/pull/47383 first

**Comment by samueljseay** [comment]
@opr The attempt in https://github.com/woocommerce/woocommerce/pull/47383 to update tests to React 18 is not going to be a quick solution as it opens up issues in other monorepo packages 😭 

We will either have to spend time continuing down the path of finishing https://github.com/woocommerce/woocommerce/pull/47383 or this PR will be blocked because there are 18 test failures.

We could skip the slotfill tests and mock the slotfill in all tests that happen to render it as an alternative. None of the solutions feel particularly good.

**Comment by samueljseay** [comment]
https://github.com/woocommerce/woocommerce/pull/47383 is looking like a viable path forward, so I'll try get that merged ahead of this PR.

**Review by opr — APPROVED** [review]
Thanks Sam, just finished another round of testing, looks great! I tested with WC Subscriptions (which fills slots in the shipping methods block and the totals meta area).

Looking great!

File: plugins/woocommerce-blocks/packages/checkout/slot/index.js
Comment (senadir) [reply plugins/woocommerce-blocks/packages/checkout/slot/index.js:19]: This was at a time when had other plans for slot fills, including publishing that package and the possibility of it not being pegged.

**Pull request merged**
